src/librustc_mir/transform/inline.rs

```diff
@@ -94,17 +94,14 @@ impl Inliner<'tcx> {
                     continue;
                 }
 
-                let self_node_id = self.tcx.hir().as_local_node_id(self.source.def_id()).unwrap();
-                let callee_node_id = self.tcx.hir().as_local_node_id(callsite.callee);
+                let callee_hir_id = self.tcx.hir().as_local_hir_id(callsite.callee);
 
-                let callee_body = if let Some(callee_node_id) = callee_node_id {
+                let callee_body = if let Some(callee_hir_id) = callee_hir_id {
+                    let self_hir_id = self.tcx.hir().as_local_hir_id(self.source.def_id()).unwrap();
                     // Avoid a cycle here by only using `optimized_mir` only if we have
-                    // a lower node id than the callee. This ensures that the callee will
-                    // not inline us. This trick only works without incremental compilation.
-                    // So don't do it if that is enabled.
-                    if !self.tcx.dep_graph.is_fully_enabled()
-                        && self_node_id.as_u32() < callee_node_id.as_u32()
-                    {
+                    // a lower span than the callee. This ensures that the callee will
+                    // not inline us.
+                    if self.tcx.hir().span(self_hir_id) < self.tcx.hir().span(callee_hir_id) {
```

Comment:
I think the result of `Span::cmp` can change without changes of the constructed MIR for a given def id. This means that it is likely not compatible with incremental compilation.

Reply:
Thanks, I reverted that bit.